src/gui/updater/appimageupdater.cpp

```diff
@@ -0,0 +1,186 @@
+            _logStatusMessages();
+            qCCritical(lcUpdater) << "Checking for update failed:" << e.what();
+            return false;
+        }
+    }
+
+    void startUpdateInBackground() noexcept
+    {
+        // monitor progress and log status messages
+        auto *timer = new QTimer(this);
+
+        timer->setInterval(100);
+
+        connect(timer, &QTimer::timeout, this, [=]() {
+            _logStatusMessages();
+
+            if (!_updater.isDone()) {
+                emit finished(!_updater.hasError());
+            }
+
+            timer->stop();
+        });
+
+        _updater.start();
+        timer->start();
+    }
+
+signals:
+    void finished(bool successfully);
+
+private:
+    appimage::update::Updater _updater;
+};
+
+} // namespace
+
+AppImageUpdater::AppImageUpdater(const QUrl &url)
+    : OCUpdater(url)
+{
+}
+
+
+bool AppImageUpdater::handleStartup()
+{
+    // nothing to do, update will be performed while app is running, if anything
+    return false;
+}
+
+void AppImageUpdater::versionInfoArrived(const UpdateInfo &info)
+{
+    const auto &currentVersion = Version::versionWithBuildNumber();
+    const auto newVersion = QVersionNumber::fromString(info.version());
+
+    if (info.version().isEmpty() || currentVersion >= newVersion) {
+        qCInfo(lcUpdater) << "Client is on latest version!";
+        setDownloadState(UpToDate);
+        return;
+    }
+
+    const auto appImageUpdaterShim = AppImageUpdaterShim::makeInstance(info.downloadUrl(), this);
+
+    if (appImageUpdaterShim == nullptr) {
+        setDownloadState(DownloadFailed);
+        return;
+    }
+
+    if (!appImageUpdaterShim->isUpdateAvailable()) {
+        qCCritical(lcUpdater) << "Update server reported that update is available, but AppImageUpdate disagrees, aborting";
+        setDownloadState(DownloadFailed);
+        return;
+    }
+
+    auto dialog = new Ui::AppImageUpdateAvailableDialog(currentVersion, newVersion, ocApp()->gui()->settingsDialog());
+
+    connect(dialog, &Ui::AppImageUpdateAvailableDialog::skipUpdateButtonClicked, this, [this]() {
+        qCInfo(lcUpdater) << "Update skipped by user";
+
+        // TODO: remember this setting
+    });
+
+    connect(dialog, &QDialog::accepted, this, [this, appImageUpdaterShim]() {
+        // binding AppImageUpdaterShim shared pointer to finished callback makes sure the updater is cleaned up when it's done
+        connect(appImageUpdaterShim, &AppImageUpdaterShim::finished, this, [this](bool succeeded) {
+            if (succeeded) {
+                qCInfo(lcUpdater) << "AppImage update complete";
+                setDownloadState(DownloadComplete);
+            } else {
+                qCInfo(lcUpdater) << "AppImage update failed";
+                setDownloadState(DownloadFailed);
+            }
+        });
+
+        setDownloadState(Downloading);
+        appImageUpdaterShim->startUpdateInBackground();
+    });
+
+    dialog->show();
```

Comment:
Please also call ownCloudGui::raiseDialog(dialog);

Reply:
The wizard raises the settings dialog rather than raising the wizard. I'd implement the same semantics in here, too. Since the dialog is modal, it'll be shown above the settings dialog then.

Reply:
Well it might do so.
We also added some hacks in there to ensure the os will bring the wizard to the front when stuff is minimized and other applications are currently active.
However a linux implementation is currently missing, which means you will just get a desktop notification that tells you that ownCloud request your attention or something like that.